src/library/dao/cratedao.cpp

```diff
@@ -18,6 +18,32 @@ CrateDAO::~CrateDAO() {
 
 void CrateDAO::initialize() {
     qDebug() << "CrateDAO::initialize()";
+
+    //get the count to allocate HashMap
+    int tracksInCratesCount = 0;
+    QSqlQuery query(m_database);
+    query.prepare("SELECT COUNT(*) from " CRATE_TRACKS_TABLE);
+
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+    }
+
+    tracksInCratesCount = query.value(0).toInt();
+
+    m_cratesTrackIsIn.reserve(tracksInCratesCount);
+
+    //now fetch all Tracks from all crates and insert them into the hashmap
+    query.prepare("SELECT track_id, crate_id from " CRATE_TRACKS_TABLE);
+    if (!query.exec()) {
+        LOG_FAILED_QUERY(query);
+    }
+
+    const int trackIdColumn = query.record().indexOf("track_id");
+    const int crateIdColumn = query.record().indexOf("crate_id");
+    while (query.next()) {
+        m_cratesTrackIsIn.insert(query.value(crateIdColumn).toInt(),
+                                 query.value(trackIdColumn).toInt());
+    }
```

Comment:
> This basically copies the whole crates table to memory. 
> Up to now we have to care about this table and the original library table. 
> This is extra work for all crate actions and a source of errors because of two data storages.

That's right -- but it's not that much memory (roughly 16-bytes per tuple so we're talking about like 16MB if every track was in a playlist in a 1-million track library). The main issue is keeping them in sync and since the DAO is already the gate-keeper it will do that fine.

I think having an in-memory index for fast lookups will help for bugs like these too:
https://bugs.launchpad.net/mixxx/+bug/1402133
https://bugs.launchpad.net/mixxx/+bug/1021422

For a GMail-style label delegate in the main library view we will need a fast lookup in memory because I can only think of 2 other ways:
* Doing an additional query per visible track in the UI.
* Incorporating a join of all crates/playlists into the root library view. 

Going with option 1 would mean terrible things for UI performance. Going with option 2 would mean adding a track to a crate would invalidate / recompute the library SQLite view which I don't think would be speedy.